File: algorithms/code/leetcode/lc168_excel_sheet_column_title/lc168_excel_sheet_column_title.py
# https://leetcode.com/problems/excel-sheet-column-title/

import unittest


class Solution:
    def convertToTitle(self, columnNumber: int) -> str:
        n = columnNumber
        buffer = ''
        condition = True
        while condition:
            n -= 1
            digit = chr(ord('A') + (n % 26))
            buffer += digit
            n -= n % 26
            n //= 26
            condition = n > 0
        return buffer[::-1]


class TestCode(unittest.TestCase):
    def test_1(self) -> None:
        assert Solution().convertToTitle(1) == 'A'

    def test_2(self) -> None:
        assert Solution().convertToTitle(2) == 'B'

    def test_3(self) -> None:
        assert Solution().convertToTitle(3) == 'C'

    def test_26(self) -> None:
        assert Solution().convertToTitle(26) == 'Z'

    def test_27(self) -> None:
        assert Solution().convertToTitle(27) == 'AA'

    def test_28(self) -> None:
        assert Solution().convertToTitle(28) == 'AB'

File: algorithms/code/leetcode/lc168_excel_sheet_column_title/lc168_excel_sheet_column_title.rs
// https://leetcode.com/problems/excel-sheet-column-title/
// #easy

pub struct Solution;

impl Solution {
    pub fn convert_to_title(mut n: i32) -> String {
        let mut buffer = String::new();
        loop {
            n -= 1;
            let digit = (b'A' + (n % 26) as u8) as char;
            buffer.push(digit);
            n -= n % 26;
            n /= 26;
            if n <= 0 {
                break;
            }
        }
        buffer.chars().rev().collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::convert_to_title(1), "A");
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::convert_to_title(2), "B");
    }

    #[test]
    fn test_3() {
        assert_eq!(Solution::convert_to_title(3), "C");
    }

    #[test]
    fn test_26() {
        assert_eq!(Solution::convert_to_title(26), "Z");
    }

    #[test]
    fn test_27() {
        assert_eq!(Solution::convert_to_title(27), "AA");
    }

    #[test]
    fn test_28() {
        assert_eq!(Solution::convert_to_title(28), "AB");
    }
}
